elfi/methods.py

```diff
@@ -135,34 +162,39 @@ def sample(self, n_samples, quantile=0.01, threshold=None):
         if quantile <= 0 or quantile > 1:
             raise ValueError("Quantile must be in range ]0, 1].")
 
-        n_sim = int(n_samples / quantile) if threshold is None else n_samples
-
-        while True:
-            logger.info("{}: Running with {} proposals."
-                        .format(self.__class__.__name__, n_sim))
-            distances, parameters = self._get_distances(n_sim)
-            distances = distances.ravel()  # avoid unnecessary indexing
-
-            if threshold is None:  # filter with quantile
-                sorted_inds = np.argsort(distances)
-                threshold = distances[ sorted_inds[n_samples-1] ]
-                accepted = sorted_inds  # only the first n_samples in `return`
-                break
-
-            else:  # filter with predefined threshold
-                accepted = distances < threshold
-                n_accepted = sum(accepted)
+        parameters = None; distances = None; accepted = None; accept_rate = None
+
+        if threshold is None:
+            # Quantile case
+            n_sim = int(np.ceil(n_samples / quantile))
+            distances, parameters = self._acquire(n_sim)
+            sorted_inds = np.argsort(distances, axis=0)
+            threshold = distances[sorted_inds[n_samples-1]].item()
```

Comment:
I'm not sure if I understood. This get's the distance of the last accepted sample in order.